### ISSUE TITLE
[BUG] Insufficient permissions when using bind mounts with rootless Docker installation

### ISSUE THREAD

**Issue by MxtOUT** [paperless-ngx/paperless-ngx#4242]
### Description

Setting up paperless with the rootless Docker installation method and using bind mounts instead of volumes fails at `docker compose run --rm webserver createsuperuser`:

```shell
[+] Creating 3/3
 ✔ Container paperless-gotenberg-1  Created                                                                                                                                              0.3s 
 ✔ Container paperless-broker-1     Created                                                                                                                                              0.3s 
 ✔ Container paperless-tika-1       Created                                                                                                                                              0.2s 
[+] Running 3/3
 ✔ Container paperless-tika-1       Started                                                                                                                                              1.3s 
 ✔ Container paperless-gotenberg-1  Started                                                                                                                                              1.3s 
 ✔ Container paperless-broker-1     Started                                                                                                                                              1.3s 
Paperless-ngx docker container starting...
Creating directory /usr/src/paperless/data/index
mkdir: cannot create directory ‘/usr/src/paperless/data/index’: Permission denied
```

This is a follow-up issue to #4191, where a similar bug, that was encountered when using volumes, was reported.
The fix of this issue in #4225 doesn't appear to affect bind mounts though, only volumes.

I made sure that I have the proper permissions in the directory where I  bind mount directories, as an `ls -al` shows:
```
total 20
drwxr-xr-x 2 pi pi 4096 Sep 22 21:42 .
drwxr-xr-x 8 pi pi 4096 Sep 22 21:42 ..
-rw-r--r-- 1 pi pi  408 Sep 22 21:23 docker-compose.env
-rw-r--r-- 1 pi pi 2825 Sep 22 21:37 docker-compose.yml
-rw-r--r-- 1 pi pi   31 Sep 15 15:38 .env
```

`id -u pi` and `id -g pi` are both `1000` and "pi" is also the currently logged in user.

I tried to trace down the issue, but couldn't locally replicate the GitHub action that builds the Docker image due to `npm` (inside the image) not finding a Python installation.

### Steps to reproduce

1. Follow the installation steps for a rootless Docker installation using the official Docker image with files at the end of this post.
2. At `docker compose run --rm webserver createsuperuser`, the aforementioned error is being thrown.
3. Observe that the directories `consume/`, `data/`, `/export`, and `/media` are being created in the correct location with the correct permissions, but they are empty.

`docker-compose.yml` (I removed the volume declarations under `volumes` and replaced them with bind mounts under `services.webserver.volumes`. Also, I checked with the `dev` paperless image, same result):
```yml
version: "3.4"
services:
  broker:
    image: docker.io/library/redis:7
    restart: unless-stopped
    volumes:
      - redisdata:/data

  webserver:
    image: ghcr.io/paperless-ngx/paperless-ngx:latest
    user: 1000:1000
    restart: unless-stopped
    depends_on:
      - broker
      - gotenberg
      - tika
    ports:
      - "8000:8000"
    healthcheck:
      test: ["CMD", "curl", "-fs", "-S", "--max-time", "2", "http://localhost:8000"]
      interval: 30s
      timeout: 10s
      retries: 5
    volumes:
      - ./data:/usr/src/paperless/data
      - ./media:/usr/src/paperless/media
      - ./export:/usr/src/paperless/export
      - ./consume:/usr/src/paperless/consume
    env_file: docker-compose.env
    environment:
      PAPERLESS_REDIS: redis://broker:6379
      PAPERLESS_TIKA_ENABLED: 1
      PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
      PAPERLESS_TIKA_ENDPOINT: http://tika:9998

  gotenberg:
    image: docker.io/gotenberg/gotenberg:7.8
    restart: unless-stopped

    # The gotenberg chromium route is used to convert .eml files. We do not
    # want to allow external content like tracking pixels or even javascript.
    command:
      - "gotenberg"
      - "--chromium-disable-javascript=true"
      - "--chromium-allow-list=file:///tmp/.*"

  tika:
    image: ghcr.io/paperless-ngx/tika:latest
    restart: unless-stopped

volumes:
  redisdata:
```

`docker-compose.env`:
```
USERMAP_UID=1000
USERMAP_GID=1000
```

`.env`:
```
COMPOSE_PROJECT_NAME=paperless
```

### Webserver logs

```bash
Doesn't start
```

### Paperless-ngx version

1.17.4

### Host OS

Raspberry Pi `arm64`

### Installation method

Docker - official image

### Configuration changes

Changed volumes to bind mounts

**Comment by stumpylog** [comment]
Your `ls` doesn't show the data, media, etc.  Just a mostly empty directory.

**Comment by stumpylog** [comment]
Yeah, this just seems like a Docker thing.  The folders on the host should exist with the right permissions first.  Everything works fine then.

**Comment by MxtOUT** [comment]
Even if I create the directories manually before, I can reproduce the issue

**Comment by stumpylog** [comment]
I can't

**Comment by MxtOUT** [comment]
Can I provide you with more info that could help with the reproduction?

**Comment by stumpylog** [comment]
I'd suggest making sure Docker is at its most up to date.  Otherwise, I can't reproduce this after creating folders (with correct permissions) before starting.

**Comment by MxtOUT** [comment]
Are you using a rootless docker installation as described in [the official docs](https://docs.docker.com/engine/security/rootless/)?
Because as per my understanding of [this post](https://forums.docker.com/t/bind-mounting-permissions-with-user-broken-using-rootless-docker/88257), the behavior I experience is to be expected.

Unless I'm running the container as `root` (who has uid:gid 0:0), uids (and gids) are being remapped according to the contents of `/etc/subuid` (resp. `/etc/subgid`).
So in the case that `/etc/subuid` contains `pi:100000:65536`, uid 1 in the container will be mapped to uid 100000 of the host.
Consequently, the `paperless` user with uid 1000 in the container will be mapped to uid 100999 of the host, thus triggering this permission error.

My solution currently is to remove the `user: 1000:1000` from the `docker-compose.yml` and set the `USERMAP_UID` and `USERMAP_GID` to 0.
I've not yet thought about the security implications of this change, but I don't think that this opens up any new attack vectors since the container is started by an unprivileged user anyway.

**Comment by MxtOUT** [comment]
I believe that this restriction should at least be mentioned in the docs

**Comment by deutschebanksync** [comment]
what happens if you chown your directory to uid 100999?

**Comment by christf** [comment]
What happens when you chown the directory on the host to uid 100999?

**Comment by MxtOUT** [comment]
I suppose that works but then I can't access the directory from my host machine with user `pi` (UID 1000)

**Comment by BenoitAnastay** [comment]
Maybe it's something with your `umask`

**Comment by the01** [comment]
> I suppose that works but then I can't access the directory from my host machine with user `pi` (UID 1000)

what if you use gid 1000?

**Comment by christf** [comment]
> I suppose that works but then I can't access the directory from my host machine with user `pi` (UID 1000)

in that case you could use extended attributes to also allow your pi user with whatever uid it has. Of course if you can stick the pi users into the same group and use the right umask then this works even without xattr. In any case, this does not seem to be a paperless-ngx issue at all but rather something around the permission scheme of file systems on unix.

**Comment by Royal2000H** [comment]
Experiencing this issue as well. Can run as root but not as rootless. Also using bind mounts.

**Comment by GuillaumeHullin** [comment]
If that can help I do use bind mount in the following way...

```bash
# create group
groupadd paperless

# create user 
useradd --shell /bin/bash --home /home/paperless --create-home --comment paperless --password \'*\' --gid paperless paperless

# change your folder permission
chmod 755 /your/paperless/folder

# here are your variable to put in your docker-compose.yml
echo "USERMAP_UID=$(id -u paperless)"
echo "USERMAP_GID=$(getent group paperless | cut -d: -f3)"
```

**Comment by srwareham** [comment]
The official docs say to run the compose file using the `user` docker compose parameter. However, doesn't the entrypoint script already handle mapping the consume/export/media files to the UID/GID specified by `USERMAP_UID` `USERMAP_GID`. My terminoloigy might be wrong here, but I think the assigned `user:1234` or whichever GID doesn't have the permissions to use `groupmod` etc and therefore can't execute the entrypoint script's permissions management.

I was trying to debug this last night and got my system working as expected by just populating `USERMAP_UID` and `USERMAP_GID` and omitting the `user` parameter.

Might just be worth editing the docs to remove the `user` parameter language? Or are there some services that are not respecting the USERMAP parameters? If so, maybe the alternative would be to amend the entrypoint script such that it doesn't try to switch users if USERMAP_UID = the passed `user` parameter?

**Comment by n00b12345** [comment]
Experiencing this issue too. Can be run as root but not as rootless.

**Comment by christf** [comment]
Why stalebot is harmful: https://drewdevault.com/2021/10/26/stalebot.html

**Comment by shamoon** [comment]
Fair enough. Since its [not even clear](https://github.com/paperless-ngx/paperless-ngx/issues/4242#issuecomment-1780092224) from the above if this is something that needs to be fixed in this software, we can move this to a discussion for now and it can be discussed free from the bother of stalebot
